core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/MergeProjectWithValues.java

```diff
@@ -0,0 +1,201 @@
+
+        // handle projection which prunes all symbols
+        if (node.getOutputSymbols().isEmpty()) {
+            return Result.ofPlanNode(new ValuesNode(valuesNode.getId(), valuesNode.getRowCount()));
+        }
+
+        // fix iteration order over ProjectNode's assignments
+        Map<Symbol, Expression> assignments = ImmutableMap.copyOf(node.getAssignments().getMap());
+
+        // handle values with no output symbols
+        if (valuesNode.getOutputSymbols().isEmpty()) {
+            ImmutableList.Builder<Expression> projectedRows = ImmutableList.builder();
+            for (int i = 0; i < valuesNode.getRowCount(); i++) {
+                projectedRows.add(new Row(ImmutableList.copyOf(assignments.values())));
+            }
+            return Result.ofPlanNode(new ValuesNode(valuesNode.getId(), node.getOutputSymbols(), projectedRows.build()));
+        }
+
+        // do not proceed if ValuesNode contains a non-deterministic expression and it is referenced more than once by the projection
+        Set<Symbol> nonDeterministicValuesOutputs = new HashSet<>();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            Row row = (Row) rowExpression;
+            for (int i = 0; i < valuesNode.getOutputSymbols().size(); i++) {
+                if (!isDeterministic(row.getItems().get(i), metadata)) {
+                    nonDeterministicValuesOutputs.add(valuesNode.getOutputSymbols().get(i));
+                }
+            }
+        }
+        Set<Symbol> multipleReferencedSymbols = node.getAssignments().getExpressions().stream()
+                .map(SymbolsExtractor::extractAll)
+                .flatMap(Collection::stream)
+                .collect(toImmutableMultiset()).entrySet().stream()
+                .filter(entry -> entry.getCount() > 1)
+                .map(Multiset.Entry::getElement)
+                .collect(toImmutableSet());
+        if (!Sets.intersection(nonDeterministicValuesOutputs, multipleReferencedSymbols).isEmpty()) {
+            return Result.empty();
+        }
+
+        // inline values expressions into projection's assignments
+        RowType projectedRowType = RowType.anonymous(assignments.keySet().stream()
+                .map(symbol -> context.getSymbolAllocator().getTypes().get(symbol))
+                .collect(toImmutableList()));
+        ImmutableList.Builder<Expression> projectedRows = ImmutableList.builder();
+        for (Expression rowExpression : valuesNode.getRows().get()) {
+            ImmutableMap.Builder<SymbolReference, Expression> mapping = ImmutableMap.builder();
+            Row row = (Row) rowExpression;
+            for (int i = 0; i < row.getItems().size(); i++) {
+                mapping.put(valuesNode.getOutputSymbols().get(i).toSymbolReference(), row.getItems().get(i));
+            }
+            Row projectedRow = new Row(assignments.values().stream()
+                    .map(expression -> replaceExpression(expression, mapping.build()))
```

Comment:
`mapping.build()` will be evaluated for each element, so move it outside.

In fact, I would move the logic for creating the mapping to a separate method altogether:

```java
private Map<SymbolReference, Expression> buildMappings(List<Symbol> symbols, Row row)
{
    ImmutableMap.Builder<SymbolReference, Expression> mappingBuilder = ImmutableMap.builder();
    for (int i = 0; i < row.getItems().size(); i++) {
        mappingBuilder.put(symbols.get(i).toSymbolReference(), row.getItems().get(i));
    }
    return mappingBuilder.build();
}
```

And then simply do:


```java
Map<SymbolReference, Expression> mappings = buildMappings(valuesNode.getOutputSymbols(), (Row) rowExpression);
Row projectedRow = new Row(assignments.values().stream()
        .map(expression -> replaceExpression(expression, mappings))
        .collect(toImmutableList()));
```